File: software/examples/delphi/ExampleCallback.pas
program ExampleCallback;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletLaserRangeFinder;

type
  TExample = class
  private
    ipcon: TIPConnection;
    lrf: TBrickletLaserRangeFinder;
  public
    procedure DistanceCB(sender: TBrickletLaserRangeFinder; const distance: word);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'XYZ'; { Change to your UID }

var
  e: TExample;

{ Callback function for distance callback (parameter has unit cm) }
procedure TExample.DistanceCB(sender: TBrickletLaserRangeFinder; const distance: word);
begin
  WriteLn(Format('Distance: %d cm', [distance]));
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  lrf := TBrickletLaserRangeFinder.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Turn laser on and wait 250ms for very first measurement to be ready }
  lrf.EnableLaser;
  Sleep(250);

  { Set Period for distance callback to 1s (1000ms)
    Note: The distance callback is only called every second if the
          distance has changed since the last call! }
  lrf.SetDistanceCallbackPeriod(1000);

  { Register distance callback to procedure DistanceCB }
  lrf.OnDistance := {$ifdef FPC}@{$endif}DistanceCB;

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/vbnet/ExampleCallback.vb
Imports System
Imports System.Threading
Imports Tinkerforge

Module ExampleCallback
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change XYZ to the UID of your Laser Range Finder Bricklet

    ' Callback subroutine for distance callback (parameter has unit cm)
    Sub DistanceCB(ByVal sender As BrickletLaserRangeFinder, ByVal distance As Integer)
        Console.WriteLine("Distance: " + distance.ToString() + " cm")
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim lrf As New BrickletLaserRangeFinder(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Turn laser on and wait 250ms for very first measurement to be ready
        lrf.EnableLaser()
        Thread.Sleep(250)

        ' Register distance callback to subroutine DistanceCB
        AddHandler lrf.Distance, AddressOf DistanceCB

        ' Set period for distance callback to 0.2s (200ms)
        ' Note: The distance callback is only called every 0.2 seconds
        '       if the distance has changed since the last call!
        lrf.SetDistanceCallbackPeriod(200)

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        lrf.DisableLaser() ' Turn laser off
        ipcon.Disconnect()
    End Sub
End Module
